igwn_auth_utils/requests.py

```diff
@@ -0,0 +1,289 @@
+# -*- coding: utf-8 -*-
+# Copyright 2021 Cardiff University
+# Distributed under the terms of the BSD-3-Clause license
+
+"""Python Requests interface with IGWN authentication
+
+This is heavily inspired by Leo Singer's excellent `requests-gracedb` package.
+"""
+
+__author__ = "Duncan Macleod <duncan.macleod@ligo.org>"
+__credits__ = "Leo Singer <leo.singer@ligo.org>"
+
+from os import getenv
+from netrc import netrc
+from urllib.parse import urlparse
+
+import requests
+
+from . import (
+    find_x509_credentials,
+    find_scitoken,
+    scitoken_authorization_header,
+)
+
+_auth_session_parameters = """
+    Parameters
+    ----------
+    cert : `str`, `tuple`, `bool`, optional
+        X.509 credential input, one of
+
+        - path to a PEM-format certificate file,
+        - a ``(cert, key)`` `tuple`,
+        - `False`: disable using X.509 completely
+        - `True`: discover a valid cert via
+          :func:`igwn_auth_utils.find_x509_credentials` and
+          error if something goes wrong
+        - `None`: try and discover a valid cert, but
+          try something else if that fails
+
+    token : `scitokens.SciToken`, `str`, `bool`, optional
+        bearer token (scitoken) input, one of
+
+        - an authorisation bearer token (`scitokens.SciToken`),
+        - a serialised token (`str`, `bytes`),
+        - `False`: disable using tokens completely
+        - `True`: discover a valid token via
+          :func:`igwn_auth_utils.find_scitoken` and
+          error if something goes wrong
+        - `None`: try and discover a valid token, but
+          try something else if that fails
+
+    token_audience, token_scope : `str`
+        the ``audience`` and ``scope`` to pass to
+        `igwn_auth_utils.find_scitoken` when discovering available tokens.
+
+    force_noauth : `bool`, optional
+        disable the use of any authorisation credentials (mainly for testing).
+
+    fail_if_noauth : `bool`, optional
+        raise an `Exception` if no authorisation credentials are presented
+        or discovered.
+
+    Raises
+    ------
+    ValueError
+        if ``fail_if_noauth=True`` is given and no authorisation
+        token/credentials are presented or discovered.
+
+    RuntimeError
+        if ``cert=True`` or ``token=True`` is given and the relevant
+        credential was not actually discovered.
+
+    Note
+    ----
+    This class requires `Requests <https://requests.readthedocs.io/>`__.
+
+    See also
+    --------
+    requests.Session
+        for details of the standard options
+    """.strip()
+
+
+class SessionAuthMixin:
+    """Mixin for :class:`requests.Session` to add support for IGWN auth.
+
+    By default this mixin will automatically attempt to discover/configure
+    a bearer token (scitoken) or an X.509 credential, with options to
+    require/disable either of those, or all authentication entirely.
+
+    {parameters}
+    """
+    def __init__(
```

Comment:
Like my `SessionAuthMixin` class, this should take an optional username and password for basic auth.

Reply:
This still doesn't have a username and password option.

Reply:
The `auth` option is a pre-existing space for a `(username, password)` tuple provided by the parent `requests.Session` object. The docstring for this code (now) clearly states that `auth`  is the correct keyword to use to pass in username/password. I would rather not overload the API with new keywords that provide the same functionality as existing ones.